lib/Model/Binary.cpp

```diff
@@ -100,6 +100,44 @@ bool Binary::verifyTypes(bool Assert) const {
   return verifyTypes(VH);
 }
 
+static bool beginsWithReservedPrefix(StringRef Name) {
+  return Name.startswith("unnamed_") or Name.startswith("revng_");
+}
+
+static bool hasReservedCustomName(const model::Type &T) {
+  StringRef CustomName = "";
+  switch (T.Kind) {
+
+  case model::TypeKind::Primitive:
+  case model::TypeKind::Invalid:
+  case model::TypeKind::Count:
+  default: {
+    // These kind of types don't have custom names, they cannot be wrong
+  } break;
+
+  case model::TypeKind::Enum: {
+    CustomName = cast<model::EnumType>(T).CustomName;
+  } break;
+  case model::TypeKind::Typedef: {
+    CustomName = cast<model::TypedefType>(T).CustomName;
+  } break;
+  case model::TypeKind::Struct: {
+    CustomName = cast<model::StructType>(T).CustomName;
+  } break;
+  case model::TypeKind::Union: {
+    CustomName = cast<model::UnionType>(T).CustomName;
+  } break;
+  case model::TypeKind::CABIFunctionType: {
+    CustomName = cast<model::CABIFunctionType>(T).CustomName;
+  } break;
+  case model::TypeKind::RawFunctionType: {
+    CustomName = cast<model::RawFunctionType>(T).CustomName;
```

Comment:
I thought, I had commited this. Sorry I'll fix this straight away.

Reply:
Uh, no. I was wrong. I tried this and it does not compile, because `PrimitiveType` does not have `CustomName`, so when `upcast` is expanded and unrolled the compilation fails.

Reply:
I also tried to tweak your `polyswitch` implementation you have around, but I couldn't get it to do what I wanted without bad modifications that would make it much less generic than it was.

Reply:
Ah yeah, you should push the switch in the lambda. But I guess there's not such a things as `switch constexpr`.